security-proxy/src/main/java/org/georchestra/security/LdapUserDetailsRequestHeaderProvider.java

```diff
@@ -46,156 +55,220 @@
 import org.springframework.security.core.Authentication;
 import org.springframework.security.core.context.SecurityContextHolder;
 import org.springframework.security.ldap.search.FilterBasedLdapUserSearch;
-import org.springframework.security.ldap.search.LdapUserSearch;
 import org.springframework.util.Assert;
 
+import com.google.common.base.Predicates;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.Maps;
+
 /**
  * Reads information from a user node in LDAP and adds the information as
  * headers to the request.
- *
+ * <p>
+ * Adds the remaining standard request headers {@code sec-org} and
+ * {@code sec-orgname} (while {@link SecurityRequestHeaderProvider} adds
+ * {@code sec-username} and {@code sec-roles}), and then any extra request
+ * header that can be extracted from the user's LDAP info and is configured in
+ * the datadirectory's {@code security-proxy/headers-mapping.properties} file.
+ * 
  * @author jeichar
+ * @see SecurityRequestHeaderProvider
  */
 public class LdapUserDetailsRequestHeaderProvider extends HeaderProvider {
 
+    private static final String CACHED_USERNAME_KEY = "security-proxy-cached-username";
+
+    private static final String CACHED_HEADERS_KEY = "security-proxy-cached-attrs";
+
     protected static final Log logger = LogFactory
             .getLog(LdapUserDetailsRequestHeaderProvider.class.getPackage().getName());
 
-    private LdapUserSearch _userSearch;
-    private Map<String, String> _headerMapping;
-    private Pattern pattern;
-    private String orgSearchBaseDN;
+    private final FilterBasedLdapUserSearch _userSearch;
+    private final Pattern orgSeachMemberOfPattern;
+    private final String orgSearchBaseDN;
+
+    private Map<String, String> _headerMapping = ImmutableMap.of();
 
     @Autowired
     private LdapTemplate ldapTemplate;
 
     @Autowired
     private GeorchestraConfiguration georchestraConfiguration;
 
-    public LdapUserDetailsRequestHeaderProvider(LdapUserSearch userSearch, String orgSearchBaseDN,
-            Map<String, String> headerMapping) {
+    public LdapUserDetailsRequestHeaderProvider(FilterBasedLdapUserSearch userSearch, String orgSearchBaseDN) {
         Assert.notNull(userSearch, "userSearch must not be null");
-        Assert.notNull(headerMapping, "headerMapping must not be null");
+        // Q: is it ok for orgSearchBaseDN to be null?
```

Comment:
Don't think so... cc @pmauduit

Reply:
if it's null, my comprehension of the next lines of code would be to have a more global regex pattern (e.g. "member == uid=blah,.*" instead of "member == uid=blah,ou=orgs,dc=geochestra,dc=org.*") so it does not matter that much ?

Reply:
that would end up in `Pattern.compile("([^=,]+)=([^=,]+),null.*")`